### PR TITLE
[FEM] implement centrifugal body loads, ccx label CENTRIF from *DLOAD…

### PR DESCRIPTION
**Pull request by UR-0**
… card

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by UR-0** [comment]
forum: [https://forum.freecadweb.org/viewtopic.php?f=18&t=57770&p=496520#p496517](url)

**Comment by luzpaz** [comment]
URL was wrong, should be https://forum.freecadweb.org/viewtopic.php?f=18&t=57770&p=496520#p496517

File: src/Mod/Fem/femobjects/constraint_centrif.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# ***************************************************************************
+# *   Copyright (c) 2020 Bernd Hahnebach <bernd@bimstatik.org>              *
```
Comment (donovaly) [inline src/Mod/Fem/femobjects/constraint_centrif.py:2]: This is not Bernd's code, but yours.

File: src/Mod/Fem/femobjects/constraint_centrif.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# *                                                                         *
+# *   You should have received a copy of the GNU Library General Public     *
+# *   License along with this program; if not, write to the Free Software   *
+# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+# *   USA                                                                   *
+# *                                                                         *
+# ***************************************************************************
+
+__title__ = "FreeCAD FEM constraint centrif document object"
+__author__ = "Bernd Hahnebach"
```
Comment (donovaly) [inline src/Mod/Fem/femobjects/constraint_centrif.py:25]: same here

File: src/Mod/Fem/femtaskpanels/task_constraint_centrif.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+# ***************************************************************************
+# *   Copyright (c) 2020 Bernd Hahnebach <bernd@bimstatik.org>              *
```
Comment (donovaly) [inline src/Mod/Fem/femtaskpanels/task_constraint_centrif.py:2]: wrong author

File: src/Mod/Fem/femsolver/calculix/writer.py [hunk 4]
```diff
@@ -909,6 +912,49 @@ def write_constraints_selfweight(self, f):
...
+        if not self.centrif_objects:
+            return
+
+        # write constraint to file
+        f.write("\n***********************************************************\n")
+        f.write("** Centrif Constraint\n")
+        f.write("** written by {} function\n".format(sys._getframe().f_code.co_name))
+        for femobj in self.centrif_objects:
+            # femobj --> dict, FreeCAD document object is femobj["Object"]
+            centrif_obj = femobj["Object"]
```
Comment (donovaly) [inline src/Mod/Fem/femsolver/calculix/writer.py:927]: Can you please document the code a bit. When had a look at the FEM code it was hard to understand it due to lack of comments.
It is not necessary to comment what you do but why. So why is the CCX code as it is?

File: src/Mod/Fem/femsolver/calculix/writer.py [hunk 5]
```diff
@@ -909,6 +912,49 @@ def write_constraints_selfweight(self, f):
...
+                direction = axiscopy.Curve.Direction
+                location = axiscopy.Curve.Location
+            else:             # no line found, set default
+                location = FreeCAD.Vector(0., 0., 0.)
+                direction = FreeCAD.Vector(0., 0., 1.)
+
+            f.write("** " + centrif_obj.Label + "\n")
+            f.write("*DLOAD\n")
+            f.write(
+                "{},CENTRIF,{:.13G},{:.13G},{:.13G},{:.13G},{:.13G},{:.13G},{:.13G}\n"
```
Comment (donovaly) [inline src/Mod/Fem/femsolver/calculix/writer.py:944]: This is an example where I don't understand why the code is as it is.

File: src/Mod/Fem/femtaskpanels/task_constraint_centrif.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+# *                                                                         *
+# *   You should have received a copy of the GNU Library General Public     *
+# *   License along with this program; if not, write to the Free Software   *
+# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+# *   USA                                                                   *
+# *                                                                         *
+# ***************************************************************************
+
+__title__ = "FreeCAD FEM constraint centrif task panel for the document object"
+__author__ = "Bernd Hahnebach"
```
Comment (donovaly) [inline src/Mod/Fem/femtaskpanels/task_constraint_centrif.py:25]: same here

File: src/Mod/Fem/femviewprovider/view_constraint_centrif.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# ***************************************************************************
+# *   Copyright (c) 2020 Bernd Hahnebach <bernd@bimstatik.org>              *
```
Comment (donovaly) [inline src/Mod/Fem/femviewprovider/view_constraint_centrif.py:2]: same here

File: src/Mod/Fem/femviewprovider/view_constraint_centrif.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+# *                                                                         *
+# *   You should have received a copy of the GNU Library General Public     *
+# *   License along with this program; if not, write to the Free Software   *
+# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+# *   USA                                                                   *
+# *                                                                         *
+# ***************************************************************************
+
+__title__ = "FreeCAD FEM constraint centrif ViewProvider for the document object"
+__author__ = "Bernd Hahnebach"
```
Comment (donovaly) [inline src/Mod/Fem/femviewprovider/view_constraint_centrif.py:25]: same here

**Review by donovaly — COMMENTED** [review]
I could only have a quick look and noticed that the author name is wrong and that the code should get more comments to understand why it is as it is.

File: src/Mod/Fem/femobjects/constraint_centrif.py (hunk 1 above)
Comment (UR-0) [reply src/Mod/Fem/femobjects/constraint_centrif.py:2]: Sorry, should improve that ;-)

File: src/Mod/Fem/femsolver/calculix/writer.py (hunk 5 above)
Comment (UR-0) [reply src/Mod/Fem/femsolver/calculix/writer.py:944]: ccx uses F20.0 FORTRAN input fields, please see in dload.f in ccx's source.
A discussion can be found here [https://forum.freecadweb.org/viewtopic.php?f=18&t=22759&#p176578](url) 
So please try "{:.13G}".format(math.sqrt(2.)*-1e100) and count char yourself.
I will add a comment

Comment (donovaly) [reply src/Mod/Fem/femsolver/calculix/writer.py:944]: This is what I mean. Another developer without this special knowledge should be able to understand the code. For example for me it was not clear that you write FORTRAN code etc.

But also the pure internal python code should get some more comments. When I started with FC I had a lot of troubles to understand the code. It was doable but costs much more time than having here and there some comments.

**Comment by luzpaz** [comment]
Any progress on this?

**Comment by luzpaz** [comment]
@UR-0 ping

**Comment by berndhahnebach** [comment]
added it to my rebase branches. https://github.com/berndhahnebach/FreeCAD_bhb/commits/femcentrif I will have a look

File: src/Mod/Fem/femsolver/calculix/writer.py (hunk 5 above)
Comment (berndhahnebach) [reply src/Mod/Fem/femsolver/calculix/writer.py:944]: link in the regard of fortran is wrong: https://forum.freecadweb.org/viewtopic.php?f=18&t=22759&#p176578

Comment (berndhahnebach) [reply src/Mod/Fem/femsolver/calculix/writer.py:944]: https://github.com/berndhahnebach/FreeCAD_bhb/commit/00af280c239

File: src/Mod/Fem/femsolver/calculix/writer.py (hunk 4 above)
Comment (berndhahnebach) [reply src/Mod/Fem/femsolver/calculix/writer.py:927]: generally this is true, but ... If a new object is added to FEM on a lot of places in the FEM code just some lines needs to be copied. I might should create some better comments in this regard. But my version of a good documentation is a good git history of new code added to FEM.

**Comment by berndhahnebach** [comment]
I have problems with the example if the mesh does not have mesh groups. See forum ...

**Comment by berndhahnebach** [comment]
https://github.com/berndhahnebach/FreeCAD_bhb/commits/femcentrif should work without groups too

**Comment by berndhahnebach** [comment]
https://forum.freecadweb.org/viewtopic.php?f=18&t=57770&p=514658#p514658

**Comment by berndhahnebach** [comment]
has been merged, but with some changes ...

**Pull request closed without merging**